\documentclass[aps,rmp,reprint,superscriptaddress,notitlepage,10pt,onecolumn]{revtex4-1}
\usepackage[utf8x]{inputenc}
\usepackage{amsmath,amsthm,amsfonts,amssymb,amscd}
\usepackage{graphicx}
\usepackage{wrapfig}
\usepackage{enumerate}
\usepackage{xr} % reference to other external latex documents
\usepackage[final]{hyperref}

\newcommand{\avg}[1]{\langle #1 \rangle}
\newcommand{\Lcore}{L_\text{core}}
\newcommand{\Lpang}{L_\text{pang}}
\newcommand{\Lgen}{L_\text{gen}}
\newcommand{\dcore}{\langle d_\text{core} \rangle}

\graphicspath{{../figs_suppl/}{./figs_suppl}}

\begin{document}
\title{PanGraph: scalable bacterial pan-genome graph construction \\ Supplementary Information}
\author{Nicholas Noll}
\affiliation{Kavli Institute for Theoretical Physics, University of California, Santa Barbara \looseness=-5}
\author{Marco Molari}
\affiliation{Swiss Institute of Bioinformatics, Basel, Switzerland \looseness=-5}
\affiliation{Biozentrum, University of Basel, Basel, Switzerland \looseness=-5}
\author{Richard A.~Neher}
\affiliation{Swiss Institute of Bioinformatics, Basel, Switzerland \looseness=-5}
\affiliation{Biozentrum, University of Basel, Basel, Switzerland \looseness=-5}

\maketitle

\section{Generation of synthetic data}
\label{supp:data_generation}

As a first test we benchmark the performances and accuracy of \textit{PanGraph} on synthetic data, obtained by simulating the evolution of a population in which individuals accumulate mutations, can loose part of the sequence from deletions, or gain new genetic material via Horizontal Gene Transfer (HGT) from other members of the population.\\
In the simulation a population of size $N$ is evolved for $T$ generations using a Wright-Fisher model \cite{hudson2002generating}. Each individual in the population has a circular genome of initial size $L$. At each generation single-nucleotide mutations occur at a rate $\mu$ per position per generation.\\
Deletions can occur at a rate $d$ per genome per generation. When a deletion is suggested on isolate $n$, having a genome length $L_n$, a new desired length $L'_n$ is extracted from a Gaussian distribution with mean $L$ and variance $\sigma^2$. If $\Delta = L'_n - L_n < 0$ then a random chunk of length $|\Delta|$ is removed from the genome, reducing it to size $L'_n$. If $\Delta > 0$ no deletion is performed. This ensures that the length distribution of genomes in the population remains close to the desired length $L$, with $\sigma^2$ being a proxy for the variance.
HGT occurs at a rate $h$ per genome per generation. Similarly to deletions, when a HGT event is suggested a new desired length $L'_n$ is extracted from the same Gaussian distribution, and this time the event is performed only if $\Delta = L'_n - L_n > 0$. In this case a random chunk of sequence of length $\Delta$ is extracted from a random individual in the parent population (possibly the parent of isolate $n$ itself) and inserted in a random position of the genome of isolate $n$.
Finally, inversions occur at a rate $i$ per genome per generation. When an inversion is proposed a random length $\Delta$ is chosen in the same way, and a random chunk of length $|\Delta|$ is inverted.
Standard values of the simulation parameters are reported in Table~\ref{table:sim-params}.

\begin{table}[hb]
    \begin{tabular}{c l l}
        \hline\hline
        Parameter & Description                 & Standard Value                  \\
        \hline
        $N$       & population size             & 100                             \\
        $T$       & n. of simulated generations & 50                              \\
        $L$       & average genome size         & $5 \cdot 10^5$ [bp]             \\
        $\sigma$  & s.t.d. of genome size       & $L/10$ [bp]                     \\
        $\mu$     & mutation rate               & 0.005 [1/bp $\cdot$ generation] \\
        $d$       & deletion rate               & 0.05 [1/generation]             \\
        $i$       & inversion rate              & 0.01 [1/generation]             \\
        $h$       & HGT rate                    & 0.1 [1/generation]              \\
        \hline
    \end{tabular}
    \caption{{\bf Simulation parameters}. Description of all simulation parameters used for the generation of synthetic data. Unless otherwise specified, the standard value is used.}
    \label{table:sim-params}
\end{table}

At the end of the simulation the resulting mosaic genomes are collected, along with the real underlying pangenome graph that can be used as the ground truth against which to check the performances of \textit{PanGraph}. The code to perform these simulations is shipped with \textit{PanGraph} in the \verb|generate| command.

\section{Benchmark of PanGraph on synthetic data}

We test the performance of \textit{PanGraph} on synthetic data, focusing on two aspects: the computational performances of the algorithm (time and memory requirements) as a function of the dataset size, and the accuracy of the algorithm in reconstructing the real pangenome graph of the population as a function of sequence divergence.

\subsection{Computational performance}

To test the computational performances of \textit{PanGraph} we generated data using the model described in the previous section, with standard value of the parameters but varying the average genome length $L=[1,5,10,50,100,500]$ kbp and the population size $N=[10,20,50,100,200,500,1000]$. For each pair of $N,L$ values we generated 50 different datasets. On each dataset we ran the \textit{PanGraph} \verb|build| command, using \textit{minimap2} as alignment kernel with \textit{asm20} option. We used standard value for the energy parameters $\alpha=100$, $\beta=10$. Runs were performed using 8 cores. For each run we measured the wall-time of the command, the maximum memory requirements (maximum resident size) and the average cpu percent. Results are displayed in Fig.~\ref{fig:benchmark-perf-suppl}.
Thanks to the guide tree architecture of \textit{PanGraph} the run time scales almost linearly with the number of isolates.

\begin{figure}[htb]
    \includegraphics[width=.9\textwidth]{figs_suppl/benchmark_suppl.pdf}
    \caption{{\bf Computational performance of PanGraph algorightm on synthetic data, as a function of dataset size}. We run PanGraph on artificially generated datasets consisting of a variable number of isolates with genomes of variable length (colorscale). For each condition we display the average and standard deviation over 50 different runs of: (A) algorithm wall-time, (B) maximum memory requirement and (C) average percentage of cpu used on a total of 8 cores.}
    \label{fig:benchmark-perf-suppl}
\end{figure}


\subsection{Accuracy in reconstructing the pangenome graph}

To test the accuracy of \textit{PanGraph} in reconstructing the pangenome graph of a set of isolates we generated artificial data using the procedure described in Section~\ref{supp:data_generation}. For each simulation we obtain both a set of genome sequences and the underlying real pangenome graph. This was done using standard value of the parameters, but varying the rate of HGT in the interval $h=[0.01,1]$ and the mutation rate $\mu=[0,0.01]$. For each $(h,\mu)$ pair 25 different sets of data were generated.\\
On each set of data we executed \textit{PanGraph} with values for the energy parameters $\alpha=0$ and $\beta=0$, and with three different alignment kernels:
\begin{itemize}
    \itemsep1pt
    \item \textit{minimap2} with option \textit{asm10},
    \item \textit{minimap2} with option \textit{asm20},
    \item \textit{mmseqs2},
\end{itemize}
thus obtaining three different pangenome graphs per dataset.\\
To link the mutation rate parameter $\mu$ with the average pairwise SNPs distance $\avg{d}$ between homologous segments in the population we evaluated the average pairwise distance for every pancontig in the pangenome graph with depth greater than one. For each graph we then perform the weighted average of these divergences, using as weight the length of the panconting. Finally, for every pair of parameters $(\mu,h)$ we average these numbers over the 25 different simulations. Results are displayed in Fig.~\ref{fig:snps-suppl}.\\
While the rate of HGT does not influence the average pairwise divergence of pancontigs, the choice of alignment kernel controls their average divergence. With the \textit{asm20} option, \textit{Minimap2} is able to correctly merge pangenome graphs with up to populations with $\avg{d} \sim 7\%$ in our simulations, while \textit{mmseqs2} reaches $\avg{d} \sim 10\%$ at the expense of higher processing time.\\
By performing a linear fit on the datapoints with $\mu \leq 0.002$ we are able to recover the conversion factor between the mutation rate and average pairwise divergence of pancontigs in our simulations: $\avg{d} \sim 20.8 \, \mu$.

\begin{figure}[htb]
    \includegraphics[width=0.5\textwidth]{figs_suppl/snps_rate_vs_divergence.pdf}
    \caption{{\bf mutation rate vs divergence on synthetic data}. We compare the mutation rate $\mu$ of our simulations with the average pairwise divergence $\avg{d}$ of sequences in pancontigs of the resulting pangenome graphs. This is done for different values of the rate of HGT $h$ and for three different alignment kernels. As expected, the divergence is only marginally influenced by the HGT rate, but the saturating value of the divergence depends strongly on the choice of alignment kernel, with \textit{mmseqs2} being able to merge sequences with average divergence higher than 10\%. To find the relationship between the mutation rate $\mu$ and the average pairwise divergence of the sequences we peform a linear fit on data points for $\mu \leq 0.002$. This provides the conversion factor $\avg{d} \sim 20.8 \, \mu$.}
    \label{fig:snps-suppl}
\end{figure}

Once this link has been established we group simulations by the value of $h$ and measure accuracy by comparing the reconstructed pangenome graphs with the ground truth provided by our simulations. In particular, for each isolate we consider the breakpoints between different pancontigs that tile the genome, and measure the displacement between the real position of these breakpoints and the position reconstructed by pangraph.\\
To evaluate this displacement it is first necessary to link breakpoints on the real and reconstructed pangenome graph. We formulate this problem as an assignment problem, and solve it numerically using the Hungarian method \cite{kuhn1955hungarian}. For each isolate we call $b_i$ for $i=1,\ldots,I$ the panconting breakpoints on the real pangenome graph, and $b_j$ for $j=1,\ldots,J$ the ones on the reconstructed pangenome graph.\footnote{To avoid artifacts generated by the default threshold distance $d=100$ bp of 
\textit{PanGraph}, if any pair of breakpoints on the same graph sits at a distance smaller than this threshold, we remove one of them.} We define a cost matrix $D_{ij} = d(b_i,b_j)$ whose elements are distances on the genome of pairs of breakpoints from the two graphs. We then numerically solve the assignment problem on this matrix, and obtain a set of $K = \min\{ I,J \}$ pairings between breakpoints such that the distance between the pairs is minimal and all the breakpoints of the graph with minimum number of breakpoints have been paired. The average distance between breakpoints in these pairs is the average breakpoint distance of the graph.\\
For each simulation we obtain a number of average breakpoint distances equal to the number of isolates in the simulation. In Fig.~\ref{fig:benchmark-accuracy-suppl} we plot the cumulative distribution of these distances, with a cutoff at 1kbp. We stratify simulations according to the average pairwise divergence of pancontigs inferred from the value of $\mu$. Each panel corresponds to a different alignment kernel. As the sequence diversity increases we observe a clear transition. For low diversity most of the breakpoints are inferred to be only a few bp away from their real position, while for highly diverged sequences the position of breakpoints is not precisely inferred and can be hundreds of bp away from their real position. For different alignment kernels this transition occurs at different values of the divergence, with \textit{mmseqs2} being the most accurate, as can be visualize from Fig.~3 in the main text.

\begin{figure}[htb]
    \includegraphics[width=\textwidth]{figs_suppl/accuracy_comparison.pdf}
    \caption{{\bf Accuracy of PanGraph algorithm on synthetic data}. For each of our generated datasets we evaluate the average displacement of inferred pancontig breakpoints. The figure represents the cumulative distribution of average misplacements, stratified by the average pairwise divergence of sequences in pancontigs (see colorbar on the right). Each panel corresponds to a different alignment kernel used to infer the pangenome graph (see panel title). As divergence increases we observe a transition in the precision of breakpoint position estimation. At low divergence the position of most breakpoints is correctly estimated withing a few bps, while at high divergence most breakpoints can be misplaced by multiple hundreds of bps. The transition value for the divergence depends on the alignemnt kernel used, with \textit{mmseqs2} having the highest accuracy at the cost of higher computational time.}
    \label{fig:benchmark-accuracy-suppl}
\end{figure}

\section{Benchmark of PanGraph on real data}

We benchmark \textit{PanGraph} on real data, focusing on the computational performances and the properties of the resulting pangenome graphs.

\subsection{Dataset selection}

We downloaded sets of annotated and complete chromosomes from RefSeq \cite{o2016reference} for 5 different bacterial species: \textit{Klebsiella Pneumoniae} (KP), \textit{Helicobacter Pylori} (HP), \textit{Prochlorococcus Marinus} (PM), \textit{Mycobacterium Tuberculosis} (MT) and \textit{Escherichia Coli} (EC). In Table~\ref{table:panx-dataset} we report some summary statistics for these datasets. The number of isolates per species ranges from several hundreds for EC to 10 for PM.
These data were previously analyzed using PanX \cite{ding2018panx}, a pipeline for the analysis of bacterial pangenome. The pipeline takes as input a set of annotated genomes, and it clusters and aligns homologous genes. This allowed us to identify core genes, and evaluate their average pairwise sequence divergence $\avg{d_\text{core}}$. We also evaluated the average genome length $
\avg{\Lgen}$, the total pangenome length $\Lpang$ and core genome length $\Lcore$. These quantities are reported in Table~\ref{table:panx-dataset}.\\
EC is the dataset with the highest number of isolates (307) and is also the one with the biggest pangenome genome. MT is the species with the smallest sequence divergence ($\dcore \sim 0.03\%$) and the biggest core genome. The two datasets with highest sequence divergence are HP ($\dcore \sim 4\%$) and PM ($\dcore \sim 24\%$). For HP the average sequence divergence sits on the limit of what \textit{PanGraph} is able to accurately merge when using the \textit{minimap2} kernel (cf. Fig.~\ref{fig:benchmark-accuracy-suppl}), but is merged correctly by the \textit{mmseqs2} kernel. The sequence divergence of PM is instead too high and cannot be handled correctly by any of the kernels.



\begin{table}[hb]
    \setlength{\tabcolsep}{9pt}
    \begin{tabular}{lccccc}
        \hline\hline
        Species         & N. isolates & $\avg{\Lgen}$ [Mbp] & $\Lpang$ [Mbp] & $\Lcore$ [Mbp] & $\dcore$ \\
        \hline
        E.~Coli         & 307         & 5.0                        & 17.7                  & 0.7                   & 1.6\%                 \\
        K.~Pneumoniae   & 109         & 5.3                        & 13.0                  & 2.3                   & 0.8\%                 \\
        H.~Pylori       & 85          & 1.6                        & 1.9                   & 0.6                   & 4.2\%                 \\
        M.~Tuberculosis & 51          & 4.4                        & 4.1                   & 2.4                   & 0.03\%                \\
        P.~Marinus      & 10          & 1.8                        & 3.3                   & 0.7                   & 26.9\%                \\
        \hline
    \end{tabular}
    \caption{{\bf Dataset overview}. We downloaded complete chromosome sequences from RefSeq for 5 different bacterial species. For each species we report the number of isolates downloaded, along with the average chromosome length $\avg{\Lgen}$, total size of pangenome  $\Lpang$ and core genome $\Lcore$, and average pairwise divergence on core genes $\dcore$.}
    \label{table:panx-dataset}
\end{table}


\subsection{Pangenome graph construction}

We built pangenome graphs using five different options for the alignment kernel:
\begin{itemize}
    \itemsep1pt
    \item kernel \textit{minimap2} with option \textit{asm10} and default values for the energy parameters $\alpha=100$, $\beta=10$.
    \item kernel \textit{minimap2} with option \textit{asm20} and default values for the energy parameters $\alpha=100$, $\beta=10$.
    \item kernel \textit{mmseqs2} and default values for the energy parameters $\alpha=100$, $\beta=10$.
    \item kernel \textit{minimap2} with option \textit{asm10} and null energy parameters $\alpha=0$, $\beta=0$.
    \item kernel \textit{mmseqs2} and null energy parameters $\alpha=0$, $\beta=0$.
\end{itemize}
From the results described in the previous section (see Fig.~\ref{fig:benchmark-accuracy-suppl}) we expect \textit{mmseqs2} to be able to merge sequences with higher divergence than \textit{minimap2}. Moreover from the definition of the pseudo-energy eq.~(1) in the main text, the value of the energy parameters $\beta$ defines an upper threshold for sequence divergence $d \sim 1/\beta$. Homologous sequences with divergence higher than this value will result in positive pseudo-energy and will not be merged in a single pancontig. For $\beta=10$ the critical divergence threshold is $d \sim 10\%$. Executing pangraph with the option $\alpha=0$, $\beta=0$ will result in the merging of more diverged sequences, at the cost of a more fragmented pangenome graph containing a higher number of shorter pancontigs.


\subsection{Benchmark results}

The results of the benchmark are displayed in Fig.~\ref{fig:panx-benchmark-suppl}. In each panel data are divided by species, with colors and texture identifying the five possible choices for the alignment kernel.\\
Execution time (cf. panel B) depends on the dataset size and ranges from few minutes for PM to several hours for EC. Aligning sequences with \textit{mmseqs2} consistently requires more time than \textit{minimap2}. The time required to build a pangenome graph for the 307 chromosomes of EC is only 5-6 h for \textit{minimap2}, and around 24h for \textit{mmseqs2}. The latter also has a higher minimum memory requirement (panel C, around 8 Gb when running on 8 cores) but this is comparable to \textit{minimap2} when aligning several tens of isolates.\\
We quantify the properties of pancontings by looking at three main statistics: their total number (panel D) the size of the minimal set of pancontigs that includes at least 50\% of all the graph sequence (L50 statistics, panel E) and the minimum length of pancontigs in this minimal set (N50 statistics, panel F). As expected, the use of null energy parameters results on average in a more fragmented graph, having more and shorter pancontigs. In general the pangenome graphs of our datasets have around $10^3$ / $10^4$ pancontigs, with 50\% of the sequence being contained in around 10\% of them, having N50 length of around 1-10 kbp.\\
The total pangenome length (panel G) depends both on the number of isolates and the ability of the alignment kernel to merge homologous sequences. When setting the energy parameters to zero more diverged sequences are merged into the same pancontigs and the total pangenome graph size is consistently smaller. The total size of the pangenome graph is in general much smaller than the cumulative size of the genomes it contains (panel I), with compression up to 1-2\% of the total size for graphs with a high number of isolates (EC) or with small accessory genome (MT). The fraction of core pancontigs (panel H) is consistent with the species expectation, with MT having the highest core fraction.\\

Two species in our datasets hit the limits of the capabilities of \textit{PanGraph}: HP and PM.\\
HP contains some homologous sequences with divergence higher than the default threshold $d = 1/\beta \sim 10\%$. Removing this threshold by using null energy parameters results in a graph with fewer pancontigs, smaller L50, greater fraction of core pangenome and higher level of sequence compression. These are all indications of more complete merging of homologous sequences. This effect is more marked for \textit{mmseqs2} than for \textit{minimap2}, in agreement with their different divergence limit (cf. Fig.~\ref{fig:benchmark-accuracy-suppl}). The HP datasets represents an example in which relaxing the default divergence threshold of pangraph and using a more accurate alignment kernel results in a more complete merging of homologous sequence.\\
Conversely, the PM dataset contains highly diverged isolates, whose divergence go beyond the capabilities of both alignment kernels ($\dcore \sim 27 \%$, cf. Table~\ref{table:panx-dataset}). Only \textit{mmseqs2} manages to partially merge some of the least diverged core sequences, but is still unable to merge the majority of the core genes. This results in a small core genome fraction and poor sequence compression. In its current state \textit{PanGraph} cannot meaningfully merge genome with such high divergence into a pangenome graph.


\begin{figure}[htb]
    \includegraphics[width=\textwidth]{figs_suppl/panx_benchmark_suppl.pdf}
    \caption{{\bf Benchmark on real data}. We run \textit{PanGraph} on datasets consisting of 5 different bacterial species, and using 5 different options for the alignment kernel as described in the text. Runs were performed using 8 cores.
        \textbf{A}: number of isolates for each species.
        \textbf{B}: algoritm wall-time.
        \textbf{C}: maximum memory usage.
        \textbf{D}: total number of pancontigs in the pangenome graph.
        \textbf{E}: minimum number of pancontigs that can add up to 50\% of the total pagenome graph length.
        \textbf{F}: threshold length such that if all pancontigs longer than this threshold are added they cover more than 50\% of the full pangenome graph.
        \textbf{G}: total length of the pangenome graph (sum of all pancontig lengths).
        \textbf{H}: fraction of the pangenome graph that is composed of core pancontigs, i.e. pancontigs that occur exactly once per isolate.
        \textbf{I}: ratio between the length of the pangenome graph and the total length of all chromosomes included in the graph.
    }
    \label{fig:panx-benchmark-suppl}
\end{figure}

% TODO: explain the meaning of L50 and N50


\section{Graph marginalization}

Given a graph composed of many different genomes, PanGraph is capable of marginalizing this graph on a subset of strains with the \verb|marginalize| command. To test this feature we built pangenome graphs on a subset of genomes from the 5 different species we previously analyzed. For each species we took 50 different strains (10 for PM)\footnote{The list of selected accession numbers for each species and pair is available on the repository: \url{https://github.com/neherlab/pangraph/blob/feat/finalize-paper/script/config/projection_strains.json}} and built an unique pangenome graph using the \verb|build| command with standard values of the parameters. For each species we then randomly picked 50 different pairs of strains (45 for PM) and extracted the corresponding sub-graphs using the \verb|marginalize| command. We compared each of these graphs to their counterpart generated simply by using the \verb|build| command on the corresponding pair of strains.\\
The comparison between the marginalized (MG) and pairwise (PW) graph consists in splitting each genome using the overlap of the breakpoints given by the two different graphs. Then for each resulting segment we check whether it is shared with the other strain of the pair on either of the graphs. This gives raise to four different possibilities, corresponding to the segment being shared (1) or not (0) on either of the two graphs (MG/PW). We consider four different classes of segments: \textit{agree on sharing} (1-1 or 0-0), \textit{shared on both} (1-1), \textit{private on both} (0-0) and \textit{disagree on sharing} (1-0 or 0-1). For each class we measure the total and average length of the segments belonging to that class, and display the results in Fig.~\ref{fig:marginalize-suppl}.\\
For all species except for PM we find that MG graphs are compatible with their PW counterpart: segments on which the graphs \textit{disagree on sharing} represent only a minor fraction of the genome (much smaller than the fraction of shared or private segments) and their size is compatible with the default block length threshold of PanGraph (100 bp). PM represents and exception, with the great majority of segments being private on each graph and having large size. This expected since for PM the strains considered are highly diverged, above the sensitivity of minimap2 ($\sim 27\%$ core divergence). This causes the majority of blocks not to be merged, and generates a strong order dependence. The size of segments with \textit{disagreement on sharing} is also the largest, some segments having size of 10s of kbp.

\begin{figure}[htb]
    \includegraphics[width=.9\textwidth]{figs_suppl/proj_fig.pdf}
    \caption{{\bf Marginalized graphs are compatible with pairwise graphs}. Comparison of marginalized (MG) and pairwise (PW) built graphs for 5 different species. For each pair of strains we measure the length of segments in which MG and PW graphs agree on whether the segments are shared on both strains (\textit{agree on sharing}). These can either be both shared (\textit{shared on both}) or both private (\textit{private on both}). We also measure the length of segments on which the MG and PW graphs disagree (\textit{disagree on sharing}). For each of these for cases we report the total length as fraction of the average genome (left column) and the average size of each segment (right column). For each category we display the distribution of average results for 50 different pairs (45 for PM).
    }
    \label{fig:marginalize-suppl}
\end{figure}

\bibliography{cite}

\end{document}